cloudinit/config/cc_ssh.py

```diff
@@ -97,6 +101,31 @@ def set_redhat_keyfile_perms(keyfile: str) -> None:
     os.chmod(f"{keyfile}.pub", permissions_public)
 
 
+def _fetch_early_keys(
+    key_names: Iterable[str], rundir: str, cfg: Config
+) -> List[str]:
+    early_keys: List[ssh_util.KeyPair] = (
+        ssh_util.wait_for_early_generated_keys(rundir)
+    )
+    if not early_keys or cfg.get("seed_random"):
```

Comment:
There is a problem with this that needs to be addressed.

Cloud-init's Azure and Openstack code has automatic entropy seeding, which bypasses cloud-config. This change highlights the fact that cloud-init implements datasource-specific code in configuration modules _and_ in datasource modules. Simply checking the merged configuration is insufficient, because for some reason when this was implemented it was decided that this shouldn't just be transformed into vendor-data to be merged (overwritten by user-data).

I strongly suspect that automatic entropy seeding exists on these platforms for legacy reasons only[1][2], and is no longer needed. However, until this tech debt has been resolved, cloud-init should still respect the entropy provided by these platforms.

[1] https://github.com/torvalds/linux/commit/f2580a907e5c0e8fc9354fd095b011301c64f949
[2] https://bugs.launchpad.net/nova/+bug/1789868

Reply:
> This if statement can just be updated to also check if random_seed is in cloud.datasource.metadata, correct?

Yes, I believe so